File: pydofus2/pydofus2/com/ankamagames/haapi/Haapi.py
import json
import ssl
from time import sleep
import httpx
from pydofus2.com.ankamagames.jerakine.logger.Logger import Logger
from pydofus2.com.ankamagames.jerakine.metaclasses.Singleton import Singleton
import cloudscraper

logger = Logger()
class HaapiException(Exception):
    pass

class Haapi(metaclass=Singleton):
    
    def __init__(self) -> None:
        self.url = "https://haapi.ankama.com"
        self.APIKEY = None
        self.login = None
        self.password = None
        self.certId = None
        self.certHash = None

    def getUrl(self, request):
        return self.url + {
            "CREATE_API_KEY": "/json/Ankama/v5/Api/CreateApiKey",
            "GET_LOGIN_TOKEN": "/json/Ankama/v5/Account/CreateToken",
        }[request]

    @property
    def clients(self):
        for client in [
            cloudscraper.create_scraper(),
            # cloudscraper.create_scraper(disableCloudflareV1=True, verify=False),
            # cloudscraper.create_scraper(interpreter='nodejs', verify=False),
            # cloudscraper.create_scraper(allow_brotli=False, verify=False),
            # httpx.Client(http2=True, verify=False),
            # httpx.Client(verify=False)
            ]:
            yield client
    
    def createAPIKEY(self, login, password, certId, certHash, game_id=102) -> str:
        logger.debug("Calling HAAPI to Create APIKEY")
        data = {
            "login": login,
            "password": password,
            "game_id": game_id,
            "long_life_token": True,
            "certificate_id": str(certId),
            "certificate_hash": str(certHash),
            "shop_key": "ZAAP",
            "payment_mode": "OK",
            "lang" : "fr"
        }
        
        nbrtries = 0
        while nbrtries < 5:
            for client in self.clients:
                    response = client.post(
                        self.getUrl("CREATE_API_KEY"),
                        data=data,
                        headers={
                            "User-Agent": "Zaap 3.6.2",
                            "Content-Type": "multipart/form-data",
                            "cache-control": "no-cache",
                        },
                    )
                    if response.headers["content-type"] == "application/json":
                        logger.debug("APIKEY created")
                        key = response.json().get("key")
                        if key:
                            self.APIKEY = key
                            return response.json()
                        else:
                            logger.debug("Error while calling HAAPI to get Login Token : %s" % response.content)
                            sleep(5)
                    else:
                        from lxml import html
                        root = html.fromstring(response.content.decode('UTF-8'))
                        error = root.xpath('//span[@class="error-description"]')[0].text
                        errorCode = root.xpath('//span[@class="code-label"]//span')[0].text
                        logger.debug(f"[Haapi error {errorCode}] : Login Token creation for login {login} failed for reason: {error}")
                        sleep(5)
            logger.debug("Failed to create APIKEY, retrying in 30 seconds")
            sleep(30)
            nbrtries += 1
        return None
            
    def regenLoginToken(self):
        return self.getLoginToken(self.login, self.password, self.certId, self.certHash)
    
    def getLoginToken(self, login, password, certId, certHash, game_id=1):
        self.login = login
        self.password = password
        self.certId = certId
        self.certHash = certHash
        logger.debug("Calling HAAPI to get Login Token")
        if not self.APIKEY:
            raise HaapiException("No haapi key found")
        nbrtries = 0
        while nbrtries < 5:
            for client in self.clients:
                response = client.get(
                    self.getUrl("GET_LOGIN_TOKEN"),
                    params={
                        "game": game_id,
                        "certificate_id": certId,
                        "certificate_hash": certHash,
                    },
                    headers={
                        "User-Agent": "Zaap1",
                        "Content-Type": "multipart/form-data",
                        "APIKEY": self.APIKEY,
                    }
                )
                logger.debug(response.content)
                if response.headers["content-type"] == "application/json":
                    token = response.json().get("token")
                    if token:
                        logger.debug("Login Token created")
                        return token
                    else:
                        logger.error("Error while calling HAAPI to get Login Token : %s" % response.json()["message"])
                        sleep(5)
                else:
                    from lxml import html
                    root = html.fromstring(response.content.decode('UTF-8'))
                    error = root.xpath('//div[@id="what-happened-section"]//p/@text')[0]
                    logger.debug("Login Token creation failed, reason: %s" % error)
                    sleep(5)
        return None

File: pyd2botApp/src/auth/Haapi.js
class RESTClient {


    HttpError = function (e, t) {
        Error.captureStackTrace(this, this.constructor), this.statusCode = e.status, this.message = "Http request failed", this.body = t
    };

    async request(url, request) {
        request.timeout = 5e3;
        request.headers = {
            "User-Agent": "Zaap 3.6.2",
            "content-type": "multipart/form-data",
        };

        console.log(request);
        process.env.DEBUG_NETWORK && console.log("[HTTP] answer", request.method, url);
        const res = await fetch(url, request);
        let textResponse = await res.text();
        if (process.env.DEBUG_NETWORK && console.log("[HTTP] result", request.method, url, res.status, res.statusText), res.status >= 400) {
            throw "application/json" === res.headers.get("content-type") && (res.body = textResponse = JSON.parse(textResponse)), new this.HttpError(textResponse);
        }
        const l = res.headers && res.headers.get("content-type") && "application/json" === res.headers.get("content-type");
        return 204 !== res.status && l && (textResponse = JSON.parse(textResponse)), res.body = textResponse, res
    }

    async get(e, t = {}) {
        return this.request(e, {
            method: "GET",
            headers: {
                "If-None-Match": null,
                ...t
            }
        })
    }

    post(url, body, headers = {}) {
        const r = headers.hasOwnProperty("Content-Type") && "application/json" === headers["Content-Type"];
        return this.request(url, {
            method: "POST",
            body: JSON.stringify(body),
            headers: headers
        })
    }

    put(e, t, n = {}) {
        const r = n.hasOwnProperty("Content-Type") && "application/json" === n["Content-Type"];
        return this.request(e, {
            method: "PUT",
            body: r ? JSON.stringify(t) : s.stringify(t),
            headers: n
        })
    }

    delete(e, t = {}) {
        return this.request(e, {
            method: "DELETE",
            headers: t
        })
    }
}


class Haapi {

    constructor() {
        this.haapiUrl = "https://haapi.ankama.com/";
        this.authHelper = require('./AuthHelper.js').instance;
        this.restClient = new RESTClient();
    }

    getUrl(requestName, body) {
        let url = this.haapiUrl + {
            ANKAMA_ACCOUNT_ACCOUNT: "json/Ankama/v5/Account/Account",
            ANKAMA_ACCOUNT_CREATE_TOKEN: "json/Ankama/v5/Account/CreateToken",
            ANKAMA_ACCOUNT_ORIGIN_WITH_API_KEY: "json/Ankama/v5/Account/OriginWithApiKey",
            ANKAMA_ACCOUNT_SEND_DEVICE_INFOS: "json/Ankama/v5/Account/SendDeviceInfos",
            ANKAMA_ACCOUNT_SEND_MAIL_VALIDATION: "json/Ankama/v5/Account/SendMailValidation",
            ANKAMA_ACCOUNT_SET_EMAIL: "json/Ankama/v5/Account/SetEmail",
            ANKAMA_ACCOUNT_SET_NICKNAME_WITH_API_KEY: "json/Ankama/v5/Account/SetNicknameWithApiKey",
            ANKAMA_ACCOUNT_SIGN_ON_WITH_API_KEY: "json/Ankama/v5/Account/SignOnWithApiKey",
            ANKAMA_ACCOUNT_SET_IDENTITY_WITH_API_KEY: "json/Ankama/v5/Account/SetIdentityWithApiKey",
            ANKAMA_ACCOUNT_STATUS: "json/Ankama/v5/Account/Status",
            ANKAMA_API_CREATE_API_KEY: "json/Ankama/v5/Api/CreateApiKey",
            ANKAMA_API_DELETE_API_KEY: "json/Ankama/v5/Api/DeleteApiKey",
            ANKAMA_API_REFRESH_API_KEY: "json/Ankama/v5/Api/RefreshApiKey",
            ANKAMA_CMS_ITEMS_GET: "json/Ankama/v5/Cms/Items/Get",
            ANKAMA_CMS_ITEMS_CAROUSEL_GET: "json/Ankama/v5/Cms/Items/Carousel/GetForLauncher",
            ANKAMA_CMS_ITEMS_GETBYID: "json/Ankama/v5/Cms/Items/GetById",
            ANKAMA_CMS_POLLINGAME_GET: "json/Ankama/v5/Cms/PollInGame/Get",
            ANKAMA_CMS_POLLINGAME_MARKASREAD: "json/Ankama/v5/Cms/PollInGame/MarkAsRead",
            ANKAMA_GAME_END_SESSION_WITH_API_KEY: "json/Ankama/v5/Game/EndSessionWithApiKey",
            ANKAMA_GAME_LIST_WITH_API_KEY: "json/Ankama/v5/Game/ListWithApiKey",
            ANKAMA_GAME_SEND_EVENTS: "json/Ankama/v5/Game/SendEvents",
            ANKAMA_GAME_START_SESSION_WITH_API_KEY: "json/Ankama/v5/Game/StartSessionWithApiKey",
            ANKAMA_LEGALS_SET_TOU_VERSION: "json/Ankama/v5/Legals/SetTouVersion",
            ANKAMA_LEGALS_TOU: "json/Ankama/v5/Legals/Tou",
            ANKAMA_MONEY_OGRINS_AMOUNT: "json/Ankama/v5/Money/OgrinsAmount",
            ANKAMA_PREMIUM_GAME_CONNECT: "json/Ankama/v5/Game/Premium/Session/Connect",
            ANKAMA_PREMIUM_GAME_DISCONNECT: "json/Ankama/v5/Game/Premium/Session/Disconnect",
            ANKAMA_PROVIDER_API_KEY_LINK: "json/Ankama/v5/Provider/ApiKeyLink",
            ANKAMA_PROVIDER_API_KEY_LOGIN: "json/Ankama/v5/Provider/ApiKeyLogin",
            ANKAMA_PROVIDER_GHOST_CREATE: "json/Ankama/v5/Provider/ApiKeyGhostCreate",
            ANKAMA_SHIELD_SECURITY_CODE: "json/Ankama/v5/Shield/SecurityCode",
            ANKAMA_SHIELD_VALIDATE_CODE: "json/Ankama/v5/Shield/ValidateCode",
            ANKAMA_SHOP_ARTICLES_LIST_BY_CATEGORY: "json/Ankama/v5/Shop/ArticlesListByCategory",
            ANKAMA_SHOP_CATEGORIES_LIST: "json/Ankama/v5/Shop/CategoriesList",
            ANKAMA_SHOP_SIMPLE_BUY: "json/Ankama/v5/Shop/SimpleBuy",
            ANKAMA_SHOP_ARTICLE_LIST_BY_ID: "json/Ankama/v5/Shop/ArticlesListByIds",
            ANKAMA_VOD_ACCESS_TOKEN_GET: "json/Ankama/v5/Vod/AccessToken/GetAccessToken"
        }[requestName];
        const s = JSON.stringify(body);
        return s && (url += "?" + s), url
    }

    createAccountToken(e, t, n) {
        e.exports = async function (e, game_id, n, certificate_id, certificate_hash) {
            getTokenUrl = this.getUrl("ANKAMA_ACCOUNT_CREATE_TOKEN", {
                game: game_id,
                certificate_id: certificate_id,
                certificate_hash: certificate_hash
            });
            try {
                return (await RESTClient.get(getTokenUrl, { APIKEY: n.key })).body.token
            } catch (e) {
                throw i.error("Error when creating token for game", n.accountId, e), 403 === e.statusCode && await r.instance.logoutAccount(n), e
            }
        }
    }

    async getAPIKey(login, password, certId, certHash) {
        var cert = this.authHelper.getStoredCertificate(login).certificate;
        var certificate_hash = this.authHelper.generateHashFromCertif(cert);
        let url = this.getUrl("ANKAMA_API_CREATE_API_KEY");
        try {
            const {
                body: body
            } = await this.restClient.post(url, {
                login: login,
                password: password,
                game_id: 102,
                long_life_token: true,
                shop_key: "ZAAP",
                payment_mode: "OK",
                lang: "fr",
                certificate_id: cert.id,
                certificate_hash: certificate_hash
            });
            return {
                key: body.key,
                accountId: body.account_id,
                refreshToken: body.refresh_token,
                security: body.data && body.data.security_state,
                reason: body.data && body.data.security_detail
            }
        } catch (e) {
            if (e.body && e.body.reason) throw new Error({
                codeError: "haapi." + e.body.reason,
                error: e
            });
            throw e;
        }
    }
}

Haapi.getAPIKey("tarik-maj@hotmail.fr", "rmrtxha1", 126304780, "f2a5726c0581d18c92b4a2278ff4c457240e62e8a64265cef8793781c641ba8c")